### PR TITLE
24_NF_main_motor_driver

### LINKED ISSUE #21: NF_MainMotorDriver
# NF_MainMotorDriver

## Description

Engine or main motor drive with H bridge

## Checklist
- [x] Select asignee(s)
- [x] Select labels (optional)
- [x] Associate to the project

## Include feature
- [ ] Create a Branch
- [ ] Implement the feature
- [ ] Explain how the feature works
- [ ] Create a Pull Request
- [ ] Link the Pull request to the issue

## Environment

* IDE: CubeIDE
* Microcontroller: Nucleo-F446RE

### PR DESCRIPTION
**Pull request by olafhuerta97**
## Description

This PR add the driver for the L298N motor driver

Fixes #21 

## Type of change

- [x] New feature

## Checklist:

- [x] Select assignee(s)
- [x] Select labels (optional)
- [x] Associate to the project
- [x] Select reviewer(s)
- [x] Check if changes generate no new errors or warnings
- [x] Link the issue #21 

> Please make sure that author does not resolve the comments added in the PR, just another member may solve the requested changes.

- [ ] Make changes if needed

> Please make sure that author does merge the PR, another member.

- [ ] Merge the PR
- [ ] Close the PR

### CODE REVIEW

**Comment by olafhuerta97** [comment]
@JDRQ98 Could you please check this first implementation

File: SystemMicrocontroller_Motors/Core/Inc/L298N_Motor_driver.h [hunk 1]
```diff
@@ -16,6 +16,13 @@ typedef enum
   L298_ERROR    = 0x01U,
 } L298N_STD_TYPE;
 
+typedef enum
+{
+	L298N_IDLE 		= 0U,
+	L298N_FORWARD  	= 1U,
+	L298N_REVERSE  	= 1U,
```
Comment (JDRQ98) [inline SystemMicrocontroller_Motors/Core/Inc/L298N_Motor_driver.h]: I think this is a typo - L298N_REVERSE should be equal to 2U

File: SystemMicrocontroller_Motors/Core/Src/L298N_Motor_driver.c [hunk 2]
```diff
@@ -116,22 +137,37 @@ void L298N_Get_Default_Config(L298N_HandleTypeDef * L298N){
...
+	if ( L298N_FORWARD == Direction) {
 		HAL_GPIO_WritePin(L298N->DirectionPin_1_Port, L298N->DirectionPin_1, SET);
 		HAL_GPIO_WritePin(L298N->DirectionPin_2_Port, L298N->DirectionPin_2, RESET);
-	}else if (REVERSE == Direction){
+	}else if (L298N_REVERSE == Direction){
 		HAL_GPIO_WritePin(L298N->DirectionPin_1_Port, L298N->DirectionPin_1, RESET);
 		HAL_GPIO_WritePin(L298N->DirectionPin_2_Port, L298N->DirectionPin_2, SET);
 	} else {
 		/**<Do nothing */
 	}
```
Comment (JDRQ98) [inline SystemMicrocontroller_Motors/Core/Src/L298N_Motor_driver.c:169]: Perhaps add a third check for if( L298N_IDLE == Direction ) and else throw an 'unexpected input' error? As it currently is, any value of Direction other than 1 or 2 would idle the motor

File: SystemMicrocontroller_Motors/Core/Src/L298N_Motor_driver.c [hunk 3]
```diff
@@ -116,22 +137,37 @@ void L298N_Get_Default_Config(L298N_HandleTypeDef * L298N){
...
 	HAL_GPIO_WritePin(L298N->DirectionPin_1_Port, L298N->DirectionPin_1, RESET);
 	HAL_GPIO_WritePin(L298N->DirectionPin_2_Port, L298N->DirectionPin_2, RESET);
 
-	if ( FORWARD == Direction) {
+	if ( L298N_FORWARD == Direction) {
 		HAL_GPIO_WritePin(L298N->DirectionPin_1_Port, L298N->DirectionPin_1, SET);
 		HAL_GPIO_WritePin(L298N->DirectionPin_2_Port, L298N->DirectionPin_2, RESET);
-	}else if (REVERSE == Direction){
+	}else if (L298N_REVERSE == Direction){
 		HAL_GPIO_WritePin(L298N->DirectionPin_1_Port, L298N->DirectionPin_1, RESET);
```
Comment (JDRQ98) [inline SystemMicrocontroller_Motors/Core/Src/L298N_Motor_driver.c:162]: > HAL_GPIO_WritePin(L298N->DirectionPin_1_Port, L298N->DirectionPin_1, RESET)

The RESET commands inside these two if statements are redundant, they run twice for the same pin

**Review by JDRQ98 — APPROVED** [review]

**Pull request merged**